keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,502 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
+it should optimize traffic by always routing to a local endpoint _when
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: ../2433-topology-aware-hints/
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+```
+<<[UNRESOLVED deprecate-iTP]>>
+
+- Deprecate `internalTrafficPolicy`? It's clear that the DNS use case
+  given in the Internal Traffic Policy KEP is not actually a good use
+  case for Internal Traffic Policy, because no one wants the behavior
+  of "I'd rather have DNS requests get dropped than have them go to
+  another node". But without the DNS use case, it's not clear that
+  there's really a strong argument for Internal Traffic Policy at all.
+
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Providing any improvements over `externalTrafficPolicy: Local` for
+  LoadBalancer services. Consensus is that [Proxy Terminating
+  Endpoints] should solve the problems that made
+  `externalTrafficPolicy: Local` unreliable for some cases.
+
+- Attempting to improve the endpoint selection algorithm for services
+  with a number of endpoints that is not (roughly) an integer multiple
+  of the number of nodes. If you have a service with 15 endpoints in a
+  cluster with 10 nodes, then the service could potentially be
+  optimized by making clients more likely to select a local endpoint,
+  but they can't _exclusively_ select local endpoints or the endpoints
+  on nodes with only a single endpoint would presumably get
+  overloaded. But this KEP does not attempt to figure out any new
+  behavior there.
+
+[Proxy Terminating Endpoints]: ../1669-graceful-termination-local-external-traffic-policy/
+
+## Proposal
+
+### User Stories
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
```

Comment:
Why is Node Local DNS not sufficient for this use case? Is it just the fail over if that is down? Would a solution to that problem be simpler, since we already have the happy path implemented?

Reply:
Hm... I hadn't seen that feature. But that requires you to do strange things with link-local IP addresses (having all DNS pods listen on the same fake IP so you can pass the same resolv.conf to all pods). And, well, we already have a solution for having pods be reachable by fake IPs that abstract away their exact location... Services! So I would say that Node-level topology is the simpler answer to the problem NodeLocal DNSCache was trying to solve.

Reply:
I do think a less bespoke answer would be nice, FWIW

Reply:
NodeLocal DNS also prevents the conn track table from filling up due to lingering entries from UDP traffic (which have to timeout).

Strangely all these answers didn't show up for me until just now (I have been checking this issue periodically throughout the day and there was nothing...).